### ISSUE TITLE
NCNN exporting somehow broke my yolov8 pre-trained model

### ISSUE THREAD

**Issue by Rockntt** [ultralytics/ultralytics#15624]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

Hi!

I was looking for a way to optimize my pre-trained YOLOv8 model for using it on Raspberry/Orange Pi and bumped into your article about converting it into NCNN format (https://docs.ultralytics.com/integrations/ncnn/).

I did so and something broke in my model so this is the visual output i get:

https://imgur.com/a/Z5CTyU

Please If it is not a bug, help me to cure it as my project deadline's burning :(

### Environment

Ultralytics YOLOv8.1.9 🚀 Python-3.11.0 torch-2.2.1+cpu CPU (12th Gen Intel Core(TM) i7-12700H)
Setup complete ✅ (20 CPUs, 15.7 GB RAM, 819.9/832.9 GB disk)

OS                  Windows-10-10.0.22631-SP0
Environment         Windows
Python              3.11.0
Install             pip
RAM                 15.71 GB
CPU                 12th Gen Intel Core(TM) i7-12700H
CUDA                None

matplotlib          ✅ 3.7.1>=3.3.0
numpy               ✅ 1.26.4>=1.22.2
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.10.1>=1.4.1
torch               ✅ 2.2.1>=1.8.0
torchvision         ✅ 0.17.1>=0.9.0
tqdm                ✅ 4.66.5>=4.64.0
psutil              ✅ 5.9.5
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0

### Minimal Reproducible Example

```
model = YOLO("./nano_ncnn_model")

while True:
    start_time = time.perf_counter()
    _, frame = cap.read()
    results = model.predict(frame, verbose=False, conf=0.4)
    annotated_frame = results[0].plot()
    cv2.imshow('CV', annotated_frame)
    ...
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by Y-T-G** [comment]
Can't see the image

**Comment by Rockntt** [comment]
> Can't see the image

my bad, will attach it right here

![Снимок экрана 2024-08-15 215932](https://github.com/user-attachments/assets/fd6dc9ff-8719-461d-add7-5ccfab73373e)

**Comment by Y-T-G** [comment]
Your ultralytics version is pretty old. Upgrade it.

**Comment by Rockntt** [comment]
> Your ultralytics version is pretty old. Upgrade it.

Installed 8.2.78, exported again and nothing changed.

Exporting log:
```
Ultralytics YOLOv8.2.78 🚀 Python-3.10.12 torch-2.3.1+cu121 CPU (Intel Xeon 2.00GHz)
Model summary (fused): 168 layers, 3,012,668 parameters, 0 gradients, 8.1 GFLOPs

PyTorch: starting from '/content/runs/detect/YOLOv8-arm2/weights/best.pt' with input shape (1, 3, 1024, 1024) BCHW and output shape(s) (1, 40, 21504) (6.1 MB)

TorchScript: starting export with torch 2.3.1+cu121...
TorchScript: export success ✅ 3.2s, saved as '/content/runs/detect/YOLOv8-arm2/weights/best.torchscript' (12.1 MB)

NCNN: starting export with NCNN 1.0.20240410...
NCNN: running '/usr/local/lib/python3.10/dist-packages/ultralytics/pnnx /content/runs/detect/YOLOv8-arm2/weights/best.torchscript ncnnparam=/content/runs/detect/YOLOv8-arm2/weights/best_ncnn_model/model.ncnn.param ncnnbin=/content/runs/detect/YOLOv8-arm2/weights/best_ncnn_model/model.ncnn.bin ncnnpy=/content/runs/detect/YOLOv8-arm2/weights/best_ncnn_model/model_ncnn.py pnnxparam=/content/runs/detect/YOLOv8-arm2/weights/best_ncnn_model/model.pnnx.param pnnxbin=/content/runs/detect/YOLOv8-arm2/weights/best_ncnn_model/model.pnnx.bin pnnxpy=/content/runs/detect/YOLOv8-arm2/weights/best_ncnn_model/model_pnnx.py pnnxonnx=/content/runs/detect/YOLOv8-arm2/weights/best_ncnn_model/model.pnnx.onnx fp16=0 device=cpu inputshape="[1, 3, 1024, 1024]"'
NCNN: export success ✅ 3.4s, saved as '/content/runs/detect/YOLOv8-arm2/weights/best_ncnn_model' (11.9 MB)

Export complete (8.7s)
Results saved to /content/runs/detect/YOLOv8-arm2/weights
Predict:         yolo predict task=detect model=/content/runs/detect/YOLOv8-arm2/weights/best_ncnn_model imgsz=1024  
Validate:        yolo val task=detect model=/content/runs/detect/YOLOv8-arm2/weights/best_ncnn_model imgsz=1024 data=/content/armenian_cv_2-5/data.yaml  
Visualize:       https://netron.app/
/content/runs/detect/YOLOv8-arm2/weights/best_ncnn_model
```

**Comment by Rockntt** [comment]
And also here's how I export it:
```
model = YOLO("/content/runs/detect/YOLOv8-arm2/weights/best.pt")
model.export(format="ncnn")
```

**Comment by glenn-jocher** [comment]
Thank you for sharing the export details. It appears the export process completed successfully. The issue might be related to the inference stage. Please ensure that the NCNN runtime environment is correctly set up and that the input image dimensions match the model's expected input size. For further assistance, you can refer to the [NCNN documentation](https://docs.ultralytics.com/integrations/ncnn/).

**Comment by Rockntt** [comment]
> Thank you for sharing the export details. It appears the export process completed successfully. The issue might be related to the inference stage. Please ensure that the NCNN runtime environment is correctly set up and that the input image dimensions match the model's expected input size. For further assistance, you can refer to the [NCNN documentation](https://docs.ultralytics.com/integrations/ncnn/).

Thanks for the answer! So there's now 2 questions left:

1) Is it true that I should specify input image size for working with NCNN? All was fine about different image size when I was using .pt YOLO.
2) I didn't get the key of using NCNN even after reading documentation: could I use NCNN only using YOLO() and model.predict() or I necessarily need to compile it on my own?

**Comment by Skillnoob** [comment]
1. For ncnn inference, you need to specify the same imgsz as you did while exporting. I had the same issue when exporting with different imgsz's
2. Yes, you can use NCNN with YOLO. You need to supply the folder like this: `model = YOLO('yolov8n_ncnn_model/')`

**Comment by Rockntt** [comment]
> 1. For ncnn inference, you need to specify the same imgsz as you did while exporting. I had the same issue when exporting with different imgsz's
> 2. Yes, you can use NCNN with YOLO. You need to supply the folder like this: `model = YOLO('yolov8n_ncnn_model/')`

Thanks for the answer!
So, if I have trained my YOLOv8 with 1024x1024 images, I should export it with 1024 imgsz too? And my input should be 1024px too? Or how can I export my model for using it with 640x480 input?

**Comment by Rockntt** [comment]
that's how I do exporting
```
yolo export model=nano.pt format=ncnn simplify=True int8=True data=data.yaml dynamic=True
```

**Comment by Skillnoob** [comment]
You should export with the same imgsz you trained with. If its buggy try slightly changing it.
The arguments you supplied after `format` are not valid arguments for ncnn. Please check over the [format table](https://docs.ultralytics.com/modes/export/#export-formats) for valid ncnn arguments.

**Comment by Rockntt** [comment]
> You should export with the same imgsz you trained with. If its buggy try slightly changing it. The arguments you supplied after `format` are not valid arguments for ncnn. Please check over the [format table](https://docs.ultralytics.com/modes/export/#export-formats) for valid ncnn arguments.

Ok, my bad I didn't notice valid arguments. Thank you very much! 

So will I be able to use model exported in 1024x1024 on 640x480 images?

**Comment by Skillnoob** [comment]
Yes, but exporting a model trained with a `imgsz` of 640 to a `imgsz` higher than that makes no sense, it will reduce your accuracy. You want to export at the same `imgsz` your model was trained at, the default is 640.

**Comment by Rockntt** [comment]
> Yes, but exporting a model trained with a `imgsz` of 640 to a `imgsz` higher than that makes no sense, it will reduce your accuracy. You want to export at the same `imgsz` your model was trained at, the default is 640.

Ok, I will retrain my model and come back with the result. Thanks!

**Comment by Rockntt** [comment]
Exported my model with
```
yolo export model=nano.pt format=ncnn half=True imgsz=640 data=data.yaml
```
and all works fine!
Thank you all very much!

**Comment by glenn-jocher** [comment]
Glad to hear it's working now! If you have any more questions, feel free to ask.
